httpbp/client_middlewares.go

```diff
@@ -0,0 +1,246 @@
+			// garbage collector overhead
+			newBreaker := pool.Get()
+			breaker, loaded := breakers.LoadOrStore(host, newBreaker)
+			if loaded {
+				defer pool.Put(newBreaker)
+			}
+
+			resp, err := breaker.(breakerbp.FailureRatioBreaker).Execute(func() (interface{}, error) {
+				resp, err := next.RoundTrip(req)
+				if err != nil {
+					return nil, err
+				}
+				// circuit break on any HTTP 5xx code
+				if resp.StatusCode >= http.StatusInternalServerError {
+					// read & close to ensure underlying RoundTripper
+					// (http.Transport) is able to re-use the persistent TCP
+					// connection.
+					_, _ = io.ReadAll(resp.Body)
+					return nil, resp.Body.Close()
+				}
+				return resp, nil
+			})
+			if resp != nil {
+				return resp.(*http.Response), err
+			}
+			return nil, err
+		})
+	}
+}
+
+// Retries provides a retry middleware by ensuring certain HTTP responses are
+// wrapped in errors. Retries wraps the ClientErrorWrapper middleware, e.g. if
+// you are using Retries there is no need to also use ClientErrorWrapper.
+func Retries(limit int, retryOptions ...retry.Option) ClientMiddleware {
+	if len(retryOptions) == 0 {
+		retryOptions = []retry.Option{retry.Attempts(1)}
+	}
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			err = retrybp.Do(req.Context(), func() error {
+				// include ClientErrorWrapper to ensure retry is applied for
+				// some HTTP 5xx responses
+				resp, err = ClientErrorWrapper(limit)(next).RoundTrip(req)
+				if err != nil {
+					return err
+				}
+				return nil
+			}, retryOptions...)
+			if err != nil {
+				return nil, err
+			}
+			return resp, nil
+		})
+	}
+}
+
+// MaxConcurrency is a middleware to limit the number of concurrent in-flight
+// requests at any given time by returning an error if the maximum is reached.
+func MaxConcurrency(maxConcurrency int64) ClientMiddleware {
+	var (
+		activeRequests int64
+	)
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			attemptedRequests := atomic.AddInt64(&activeRequests, 1)
+			defer atomic.AddInt64(&activeRequests, -1)
+
+			if maxConcurrency > 0 && attemptedRequests > maxConcurrency {
+				return nil, ErrConcurrencyLimit
+			}
+			return next.RoundTrip(req)
+		})
+	}
+}
+
+// MonitorClient is an HTTP client middleware that wraps HTTP requests in a
+// client span.
+func MonitorClient(slug string) ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripperFunc(func(req *http.Request) (resp *http.Response, err error) {
+			span, ctx := opentracing.StartSpanFromContext(
+				req.Context(),
+				slug+".request",
+				tracing.SpanTypeOption{Type: tracing.SpanTypeClient},
+			)
+			span.SetTag("http.method", req.Method)
+			span.SetTag("http.url", req.URL)
```

Comment:
are there any concerns about cardinality in this tag?

Reply:
this is span tag, not metrics tag (unless someone adds this into the global allow-list). there's no cardinality concerns on span tags.